source/common/stats/symbol_table_impl.h

```diff
@@ -417,81 +522,28 @@ StatName StatNameStorage::statName() const { return StatName(bytes_.get()); }
-
-  // Obtains new backing storage for an already existing StatName.
-  StatNameManagedStorage(StatName src, SymbolTable& table)
-      : StatNameStorage(src, table), symbol_table_(table) {}
-
-  ~StatNameManagedStorage() { free(symbol_table_); }
-
-  SymbolTable& symbolTable() { return symbol_table_; }
-  const SymbolTable& constSymbolTable() const { return symbol_table_; }
-
-private:
-  SymbolTable& symbol_table_;
-};
-
-/**
- * Maintains storage for a collection of StatName objects. Like
- * StatNameManagedStorage, this has an RAII usage model, taking
- * care of decrementing ref-counts in the SymbolTable for all
- * contained StatNames on destruction or on clear();
- *
- * Example usage:
- *   StatNamePool pool(symbol_table);
- *   StatName name1 = pool.add("name1");
- *   StatName name2 = pool.add("name2");
- *   uint8_t* storage = pool.addReturningStorage("name3");
- *   StatName name3(storage);
+ * per-stat overhead. Consider also use of StatNamePool or StatNameSet.
  */
-class StatNamePool {
+class StatNameManagedStorage {
 public:
-  explicit StatNamePool(SymbolTable& symbol_table) : symbol_table_(symbol_table) {}
-  ~StatNamePool() { clear(); }
+  StatNameManagedStorage(absl::string_view name, SymbolTable& table);
+  ~StatNameManagedStorage();
 
   /**
-   * Removes all StatNames from the pool.
-   */
-  void clear();
-
-  /**
-   * @param name the name to add the container.
-   * @return the StatName held in the container for this name.
-   */
-  StatName add(absl::string_view name);
-
-  /**
-   * Does essentially the same thing as add(), but returns the storage as a
-   * pointer which can later be used to create a StatName. This can be used
-   * to accumulate a vector of uint8_t* which can later be used to create
-   * StatName objects on demand.
+   * Returns stored StatName.
    *
-   * The use-case for this is in source/common/http/codes.cc, where we have a
-   * fixed sized array of atomic pointers, indexed by HTTP code. This API
-   * enables storing the allocated stat-name in that array of atomics, which
-   * enables content-avoidance when finding StatNames for frequently used HTTP
-   * codes.
-   *
-   * @param name the name to add the container.
-   * @return a pointer to the bytes held in the container for this name, suitable for
-   *         using to construct a StatName.
+   * Implementation note: We use storage_.builtin_ even if we are managing the
+   * storage in this instance, as it's the same pointer in the union either way,
+   * and the StatName treats its arg as const.
    */
-  uint8_t* addReturningStorage(absl::string_view name);
+  StatName statName() const { return StatName(storage_.builtin_); }
 
 private:
-  // We keep the stat names in a vector of StatNameStorage, storing the
-  // SymbolTable reference separately. This saves 8 bytes per StatName,
-  // at the cost of having a destructor that calls clear().
-  SymbolTable& symbol_table_;
-  std::vector<StatNameStorage> storage_vector_;
+  union {
```

Comment:
@htuch let me just point you directly to the most controversial part of this PR: this union of pointers.

I could also have avoided a union and used const_cast, but I felt like this expressed the semantics a little better.

Reply:
I think the complexity of stats is probably more controversial than this particular C++ styleism :) I'm very keen on seeing a stats fuzzer landed sooner rather than later for this reason. Here's the rationale:
1. While we have new and old versions, we can differentially fuzz and actually validate the new implementation is doing the right thing, continously as new optimizations are landed.
2. While stats might not be viewed as security critical, I actually think there are potential attacks via the control plane, e.g. if an untrusted party can name clusters or control the sequence of add/removal.

I think such a fuzzer can be written in O(1-2 hours). If we landed this, it would make it easier for review (cough).